### PR TITLE
[sw] Coverage support for OpenTitan

### PR DESCRIPTION
**Pull request by alphan**
This PR introduces coverage support for OpenTitan which enables collecting coverage data for on-device tests, or any executable compiled with coverage support. Please see below for a brief description of each commit included in this PR.

* **[sw] Vendor in llvm_clang_rt_profile**
  This change adds a vendor.hjson file and a patch to be able to build the LLVM profiling runtime support library for OpenTitan. This library relies heavily on file I/O, which we don't have, and supports multiple platforms that we are not interested in. Thus, the hjson file excludes all files that are irrelevant to OpenTitan while the patch removes the headers that we don't need and adjusts the paths of several headers.

* **Update llvm_clang_rt_profile to llvm/llvm-project@4fccdd5c85d**
  Update code from upstream repository https://github.com/llvm/llvm-project.git to revision 4fccdd5c85d05b3b03d029f533762e233259573c

* **[sw] Modify flash_link.ld for llvm coverage**
  When code is compiled with coverage enabled, the compiler emits various symbols and sections that are used at runtime to collect coverage data. This change modifies `flash_link.ld` linker script to preserve these sections.

* **[sw] Modify flash_crt.S for llvm coverage**
  When code is compiled with coverage enabled, the compiler emits pointers to functions (in `.init_array`) that must be called in order to initialize the LLVM profile buffer. This change modifies `flash_crt.S` to call these functions before the main program is executed.

* **[sw] Implement test_coverage library**
  This change:
  \- Introduces the `test_coverage` library that provides LLVM profiling runtime support and a function for sending the LLVM profile buffer over UART, and
  \- Introduces a new meson option named `coverage` to define the relevant dependencies conditionally.

* **[sw] Integrate test_coverage into test_main**
This change:
  \- Adds `test_coverage` library to the dependencies of `test_main`,  and
  \- Adds a call to `test_coverage_send_buffer` at the end of a test.
If coverage is enabled, the actual definition of `test_coverage_send_buffer` in `test_coverage_llvm` library gets linked in which sends the LLVM profile buffer over UART. If not, the NOP definition in `test_coverage_none` gets linked in.

* **[sw] Write a simple test to test coverage functionality**
  This change introduces a simple on-device test to test coverage functionality.

* **[util] Implement a script for extracting profile data from device output**
  This change introduces a python script that extracts profile data from device output.

## Workflow
Please note that this PR already adds coverage support to `test_main`, so on-device tests only need the `collect_coverage` dependency to collect coverage data.
1. Add `collect_coverage` dependency to the desired libraries.
2. `meson_init.sh -c` to enable coverage.
3. `stty -F /dev/ttyUSB0 icanon` to make sure that `EOT` is recognized and `cat` eventually exits.
3. `cat /dev/ttyUSB0 | util/device_profile_data.py foo.profraw` to capture and extract coverage data over UART.
4. Build and program device. Tests (or other executables) with coverage send the LLVM profile data over UART like this:
```
LLVM profile data (length: 312, CRC32: 370887ED):
8152666F72706CFF05000000000...8010065D01D0F000000000000
```
5. Index raw profile data: `llvm-profdata merge -sparse foo.profraw -o foo.profdata`
6. Generate coverage reports, e.g. `llvm-cov show OBJECT_FILE -instr-profile=foo.profdata`. [Sample report](https://gist.github.com/alphan/96577d7f69fc807b098eb3b4a44616b2) for `coverage_test.c`.

## Open Questions
* [X] It looks like we have to copy the entire LLVM profile buffer in a single call which means we need a relatively large buffer. We need to understand if we need to worry about this or not and how the size of this buffer scales with number of functions and lines of code.
  * Size of the profile buffer for `sha256_test` is 856 bytes when collecting coverage for `sha256_test.c`, `hw_sha256.c`, and `hmac.c`. The total size of the buffer in `test_coverage.c` is currently 8KB and we have 64KB of RAM.
* [X] Linking with multiple static libraries compiled with coverage enabled.
  * Please see above (`sha256_test`).
* [X] `gcc` builds fail.
  * Introduced a new meson option named `coverage` and a compiler check.
* [X] It looks like the profile data does not have any protection, so I thought it would be prudent to add length and CRC32. While I understand that this is not urgently needed, I think it is nice to have.
* [X] Changes in `flash_link.ld`: While I tried to do my best by looking at `objdump` output and llvm code, please take a close look at my changes. You can find `objdump` of `coverage_test.c.o` [here](https://gist.github.com/alphan/979d1ec899a278e8ab85f7656115b030) to see the sections added by the compiler when coverage is enabled (they all start with `__llvm`).
  * Marking this as done since coverage reports for `coverage_test` and `sha256_test` look OK, but we can always revisit these changes if we notice any issues.

**Remaining tasks are captured in #3155.**

## References
* https://clang.llvm.org/docs/SourceBasedCodeCoverage.html
* https://llvm.org/docs/CoverageMappingFormat.html

### CODE REVIEW

File: sw/device/exts/common/flash_link.ld
Comment (lenary) [inline sw/device/exts/common/flash_link.ld:73]: I *think* this needs to be aligned.

File: sw/device/lib/testing/test_coverage.c
Comment (lenary) [inline sw/device/lib/testing/test_coverage.c:58]: We should probably do this correctly, with a for loop between `_init_array_start` and `_init_array_end`, calling each pointer.

File: sw/device/exts/common/flash_link.ld
Comment (alphan) [reply sw/device/exts/common/flash_link.ld:73]: I think so, too. Done.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:58]: Thanks for catching this. I remember we talked about this and whether we should do this in `flash_crt.S` instead of a function. I _guess_ it is OK to do this here since we don't have anything else in `init_array` at the moment but let me know if you want me to move this to `flash_crt.S`. Done.

**Comment by alphan** [comment]
> You'll need to add the vendor directory to the exclusions in util/licence-checker.hjson in order to pass CI. This is correct because we don't want to put lowrisc headers on those files (they're from LLVM).

`util/licence-checker.hjson` already has the following
```
 12     # Exclude anything in vendored directories
 13     '*/vendor/*/*',
```
and according to the CI logs, license check passed but `clang-format` check failed due to an empty line at the end of a file. I will push the updates and keep an eye on CI. Thanks!

File: sw/device/lib/testing/meson.build
Comment (mcy) [inline sw/device/lib/testing/meson.build:33]: Nit: Given that this is an opaque dependency, why don't we just name it `enable_coverage`?

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:36]: Maybe just init this to UINT32_MAX?

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:46]: This might be slightly clearer around the control flow:
```c
for (int i  = 0; i < len; ++i) {
  crc ^= buf[i];
  for (int j = 0; j < CHAR_BITS; ++j) {
    crc >>= 1;
    if (crc & 1) { // I might != 0 this personally.
      crc ^= kPoly;
    }
  }
}
```
You don't have to do this but maybe it could help make this bit wrangling a bit clearer.

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:57]: Maybe it's worth adding a non-standard specifier to base_printf that does this, in analogy to `%z`? That way you can use a single base_printf in function below.

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:19]: C decl syntax is awful. Maybe consider writing a typedef like
```c
typedef void (*init_t)(void);
```
or something...?

File: sw/device/lib/testing/test_coverage.h
Comment (mcy) [inline sw/device/lib/testing/test_coverage.h:15]: Do you think you could expand on these comments to describe what they do in a bit more detail (i.e. mostly contractual obligations on callers).

File: sw/device/lib/testing/test_coverage.h
Comment (mcy) [inline sw/device/lib/testing/test_coverage.h:8]: Needs to be `(void)`, not `()`.

File: sw/device/tests/coverage_test.c
Comment (mcy) [inline sw/device/tests/coverage_test.c:14]: Optional: name this something like `kSpinnerChars`.

Also, shouldn't this be declared as a `char[]` rather than a `char *`?

File: sw/device/tests/coverage_test.c
Comment (mcy) [inline sw/device/tests/coverage_test.c:27]: This is the default, so you can just initialize it as `{}`.

File: sw/device/tests/coverage_test.c
Comment (mcy) [inline sw/device/tests/coverage_test.c:11]: Standard headers should go before project headers.

File: sw/device/exts/common/flash_link.ld
Comment (mcy) [inline sw/device/exts/common/flash_link.ld:77]: I would include a `. = ALIGN(4);` here.

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:13]: Is this symbol LLVM magic? Can you describe who is actually checking for this symbol?

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:64]: Honestly, this seems like the kind of code that belongs in the CRT?

Comment (luismarques) [reply sw/device/lib/testing/test_coverage.c:64]: True. I don't know how problematic it would be to move it there though, given that we presumably want to conditionally include this coverage stuff.

File: sw/device/lib/testing/test_coverage.c
Comment (luismarques) [inline sw/device/lib/testing/test_coverage.c:69]: Is this going to scale to actual programs that we want to get coverage from?

File: sw/device/lib/testing/test_coverage.c
Comment (luismarques) [inline sw/device/lib/testing/test_coverage.c:72]:
```
if (buf_size > sizeof(buf)) {
  base_printf("\nLLVM profile data is too large to handle\r\n");
  return;
}
```
Not having the normal output should already be handled by the Python extraction tool, that `printf` is just for the human to figure out what went wrong.

File: sw/device/tests/coverage_test.c
Comment (luismarques) [inline sw/device/tests/coverage_test.c:13]: The sample coverage report seemed fine, but I had wondered if we should have function attributes to minimize optimizations that could mess up the coverage report. E.g. don't inline, don't unroll loops.

File: sw/device/lib/testing/meson.build
Comment (alphan) [reply sw/device/lib/testing/meson.build:33]: Sounds good to me. Done.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:36]: (Applies to your next comment too) I implemented this as close as possible to the code in `zlib.h` to make reviewers' lives easier. Given that I didn't receive any comments regarding its correctness, I cleaned this up.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:57]: I'm also not sure if we would be able to handle length cleanly. @mcy?

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:46]: For future reference: This is actually not correct. LSB of `crc` must be checked before it is shifted right, i.e. `crc >>= 1` should not be before the if statement.

File: sw/device/tests/coverage_test.c
Comment (alphan) [reply sw/device/tests/coverage_test.c:11]: I think we need to update `.clang-format` for this, but I'm more used to local to global ordering to catch hidden dependencies in headers.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:13]: I thought the comment was descriptive enough. Please also see [this section](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#using-the-profiling-runtime-without-static-initializers) in the first reference.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:69]: That is an open question but I have one datapoint :) The length of the profile buffer for `coverage_test.c` is 312 bytes.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:72]: Good suggestion, applied with minor edits: Error message also includes the size of the LLVM buffer and an `EOT` so that `cat` can exit. Also verified that the python script handles this properly.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c]: I actually ended up moving this to `flash_crt.S`. Please let me know what you think.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:19]: Moved to `flash_crt.S`.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:64]: [I agree](https://github.com/lowRISC/opentitan/pull/3080#discussion_r467032508), moved to `flash_crt.S`.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:69]: I tried running `sha256_test` with coverage (3 TUs with coverage: `sha256_test.c`, `hw_sha256.c`, and `hmac.c`) and the buffer size was 856 bytes. There is some fixed cost per file and function but AFAICT the buffer size mostly depends on the number of [mapping regions](https://github.com/lowRISC/opentitan/pull/3080#discussion_r467032508). It looks like we won't have an issue with actual programs but nevertheless I changed the size of the buffer to 8KB to be safe (for reference, our RAM is 64KB according to `top_earlgrey_memory.ld`).

**Comment by alphan** [comment]
> Should this PR contain a check for CI to run the coverage test program and ensure that the coverage matches what was expect?

I plan to add CI support in a separate PR because this PR is already quite large and we will also need some tooling to process the raw profile data.

File: sw/device/tests/coverage_test.c
Comment (alphan) [reply sw/device/tests/coverage_test.c:13]: Great question, I'm also not sure. I think we can decide if this is needed when we start collecting coverage from actual programs and review the reports.

**Comment by alphan** [comment]
In the last push (grouped by commit):
* **[sw] Modify flash_link.ld for llvm coverage**
  * Align the end of `.init_array` section.
* **[sw] Modify flash_crt.S for llvm coverage**
  * Call function pointers in `.init_array` in `flash_crt.S`.
* **[sw] Implement test_coverage library**
  * Add false-by-default `coverage` option to meson, `meson_init.sh` flag: `-c`.
  * Define coverage dependencies based on `coverage` option, `NOP` if `coverage` is false.
  * Increase size of the buffer in `test_coverage_send_buffer` to 8KB.
  * Implement comments.
* **[sw] Integrate test_coverage into test_main**
  * Add a weak definition for `test_coverage_send_buffer` in `test_main.c`.
* **[sw] Write a simple test to test coverage functionality**
  * Implement comments.
* **[util] Implement a script for extracting profile data from device output**
  * Ignore utf-8 errors due to device resetting before programming.
  * Process lines in reverse order.
  * Run `yapf` and `pylint`.

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [reply sw/device/lib/testing/test_coverage.c]: Oh. You're right. Ugh. Personally I would still write it this way:
```
bool lsb = crc & 1;
crc >>= 1;
if (lsb) {
  crc &= kPoly;
}
```
but *shrug*.

File: sw/device/tests/coverage_test.c
Comment (mcy) [reply sw/device/tests/coverage_test.c:11]: I think just putting them in separate stanzas will make clang-format keep way from them. (Style guide actually mandates self-header, system-headers, project-headers, at any rate).

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [reply sw/device/lib/testing/test_coverage.c]: Our non-conforming `printf` provides a couple of non-standard format characters. `%z` consumes two arguments: a `size_t` and a `char*`, behaves like `%s` but with the given length. My suggestion was to add a format specifier (e.g., `%y` and `%Y`) which behave the same way, but which print hex rather than ASCII. We can punt this but it seems like such a useful formatting strategy that I thought adding it to print.h was worth it. *shrug*.

File: sw/device/exts/common/flash_crt.S
Comment (mcy) [inline sw/device/exts/common/flash_crt.S:101]: `t0` and `t1` are both caller-saved. Maybe consider using `s0` and `s1` to make the spilling the C function's problem?

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [reply sw/device/lib/testing/test_coverage.c:13]: Can you just include a link to that in this file? "The linker" is really vague in a "is GNU ld somehow figuring this out?" way.

File: sw/device/lib/testing/test_coverage.c
Comment (mcy) [inline sw/device/lib/testing/test_coverage.c:59]: I think for something like this you really aught to just use a LOG macro.

File: sw/device/exts/common/flash_crt.S
Comment (alphan) [reply sw/device/exts/common/flash_crt.S:101]: Good point! I actually read the table in the risc-v spec but forgot to update this. Done.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:59]: I agree, done.

File: sw/device/tests/coverage_test.c
Comment (alphan) [reply sw/device/tests/coverage_test.c:11]: > I think just putting them in separate stanzas will make clang-format keep way from them.

Unfortunately it doesn't.

File: sw/device/lib/testing/test_coverage.c
Comment (alphan) [reply sw/device/lib/testing/test_coverage.c]: Thanks for the explanation @mcy. I think we should definitely do this because the log library already has an efficient way for writing logs for DV. Let me do this in a separate PR once this lands to address the performance issue in DV. Updated #3155.

**Comment by alphan** [comment]
Implemented @mcy's comments.

File: sw/device/lib/testing/test_coverage.c
Comment (sriyerg) [inline sw/device/lib/testing/test_coverage.c:50]: If you replace this with LOG_INFO (we can create something specific for transmitting profile data specifically if needed), then we could get the coverage collection working in DV as well.

Comment (alphan) [reply sw/device/lib/testing/test_coverage.c:50]: Thanks @sriyerg. I agree, but I prefer not to do this _right now_ because it appends a new line after each message at runtime. We plan to add a non-standard specifier to `base_printf` as @mcy suggested to print the entire buffer in a single call (please see [this thread](https://github.com/lowRISC/opentitan/pull/3080#discussion_r471634052)) and then use `LOG_INFO` to solve this issue (tracked in #3155).

Comment (sriyerg) [reply sw/device/lib/testing/test_coverage.c:50]: Okay, SG.

File: sw/device/lib/testing/test_main.c
Comment (lenary) [inline sw/device/lib/testing/test_main.c:16]: We're trying to avoid `__attribute__((weak))`.

I think the solution here is to have two "implementations" of `test_coverage.h` - `test_coverage_llvm.c` and `test_coverage_none.c` - the second of which just has an empty function here. Thus we can always link in `test_coverage_none.o`, unless we have coverage enabled where we'd link in `test_coverage_llvm.o`. Does that make sense?

File: sw/device/tests/coverage_test.c
Comment (lenary) [reply sw/device/tests/coverage_test.c:11]: Oh weird. I can certainly make system headers come before project headers in the clang format config.

File: sw/device/tests/coverage_test.c
Comment (lenary) [inline sw/device/tests/coverage_test.c:34]: :D

File: sw/vendor/meson.build
Comment (lenary) [inline sw/vendor/meson.build:57]: Does it not work if this is not in the include directories?

File: sw/device/exts/common/flash_crt.S
Comment (alphan) [reply sw/device/exts/common/flash_crt.S]: Fixed in previous push.

File: sw/device/exts/common/flash_link.ld
Comment (alphan) [reply sw/device/exts/common/flash_link.ld:77]: Thanks for catching this. I think I misunderstood @mcy's comment assuming it was to make sure that the section ends at a word boundary since `_init_array_end` is only used in a comparison. Done.

File: meson_init.sh
Comment (alphan) [reply meson_init.sh:52]: Thanks!

File: sw/device/lib/testing/meson.build
Comment (alphan) [reply sw/device/lib/testing/meson.build:42]: ATM `compile_args` should be added as a dependency to the libraries that we want to collect coverage for. I thought using an _opt-in_ approach would be beneficial for minimizing the size of the profile buffer, e.g. we probably don't care about the coverage of the logging library when collecting coverage for `sha256_test` and its dependencies, but maybe I was overthinking this. Please let me know if you think we should do this instead.

File: sw/vendor/meson.build
Comment (alphan) [reply sw/vendor/meson.build:57]: Thanks for catching this, it does. Done.

File: sw/device/tests/coverage_test.c
Comment (alphan) [reply sw/device/tests/coverage_test.c:11]: I wonder if this is due to the version of the `clang-format` that I have. I can disable it and make this change manually before pushing my latest changes.

File: sw/device/lib/testing/test_main.c
Comment (alphan) [reply sw/device/lib/testing/test_main.c:16]: It definitely does. Apologies if we've already talked about this and I forgot about it. Done, PTAL.

**Comment by alphan** [comment]
Addressed @lenary's comments.

File: sw/device/lib/testing/meson.build
Comment (lenary) [reply sw/device/lib/testing/meson.build:47]: Ah, I see where you're coming from.

File: sw/device/tests/coverage_test.c
Comment (lenary) [reply sw/device/tests/coverage_test.c:11]: If CI is passing, don't worry about reformatting. I've got a different file with fixes to the include order in clang-format.

Comment (alphan) [reply sw/device/tests/coverage_test.c:11]: Sounds good, making this change manually makes CI unhappy as expected. I will `clang-format` and update this part.

**Review by lenary — APPROVED** [review]
Thanks, I think I'm happy with this!

@mcy can you leave an approving review if all your comments were addressed? I don't think the include ordering should be a blocker now.

**Comment by alphan** [comment]
Applied CI patch to fix the order of includes in `test_coverage_llvm.c`.

**Comment by alphan** [comment]
> Thanks, I think I'm happy with this!
> 
> @mcy can you leave an approving review if all your comments were addressed? I don't think the include ordering should be a blocker now.

Thanks @lenary!

**Comment by alphan** [comment]
> LG =)

Thanks!

**Pull request merged**
